lib/common/ipc_pacemakerd.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright 2020 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+
+#include <stdlib.h>
+#include <time.h>
+
+#include <crm/crm.h>
+#include <crm/msg_xml.h>
+#include <crm/common/xml.h>
+#include <crm/common/ipc.h>
+#include <crm/common/ipc_internal.h>
+#include <crm/common/ipc_pacemakerd.h>
+#include "crmcommon_private.h"
+
+typedef struct pacemakerd_api_private_s {
+    enum pcmk_pacemakerd_state state;
+    char *client_uuid;
+    unsigned int replies_expected;
+} pacemakerd_api_private_t;
+
+static const char *pacemakerd_state_str[] = {
+    XML_PING_ATTR_PACEMAKERDSTATE_INIT,
+    XML_PING_ATTR_PACEMAKERDSTATE_STARTINGDAEMONS,
+    XML_PING_ATTR_PACEMAKERDSTATE_WAITPING,
+    XML_PING_ATTR_PACEMAKERDSTATE_RUNNING,
+    XML_PING_ATTR_PACEMAKERDSTATE_SHUTTINGDOWN,
+    XML_PING_ATTR_PACEMAKERDSTATE_SHUTDOWNCOMPLETE
+};
+
+enum pcmk_pacemakerd_state
+pcmk_pacemakerd_api_pacemakerd_state_text2enum(const char *state)
+{
+    int i;
+
+    if (state == NULL) {
+        return pcmk_pacemakerd_state_invalid;
+    }
+    for (i=pcmk_pacemakerd_state_init; i <= pcmk_pacemakerd_state_max;
+         i++) {
+        if (crm_str_eq(state, pacemakerd_state_str[i], TRUE)) {
+            return i;
+        }
+    }
+    return pcmk_pacemakerd_state_invalid;
+}
+
+const char *
+pcmk_pacemakerd_api_pacemakerd_state_enum2text(
+    enum pcmk_pacemakerd_state state)
+{
+    if ((state >= pcmk_pacemakerd_state_init) &&
+        (state <= pcmk_pacemakerd_state_max)) {
+        return pacemakerd_state_str[state];
+    }
+    return NULL;
+}
+
+// \return Standard Pacemaker return code
+static int
+new_data(pcmk_ipc_api_t *api)
+{
+    struct pacemakerd_api_private_s *private = NULL;
+
+    api->api_data = calloc(1, sizeof(struct pacemakerd_api_private_s));
+
+    if (api->api_data == NULL) {
+        return errno;
+    }
+
+    private = api->api_data;
+    private->state = pcmk_pacemakerd_state_invalid;
+    /* other as with cib or controld pacemakerd is just addressed
+       from the local node - pid is sufficient as an ID
+     */
```

Comment:
actually that's only true of pacemakerd :)  the others can be addressed from a pacemaker remote node, so they can potentially have collisions currently

Reply:
rephrased a bit - hopefully making it more obvious